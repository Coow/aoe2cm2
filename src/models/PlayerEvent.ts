import Civilisation from "./Civilisation";
import ActionType from "../constants/ActionType";
import Player from "../constants/Player";

class PlayerEvent {
    public readonly player: Player;
    public readonly executingPlayer: Player;
    public readonly actionType: ActionType;
    public civilisation: Civilisation;
    public offset: number;

    constructor(player: Player, actionType: ActionType, civilisation: Civilisation, executingPlayer: Player = player) {
        this.player = player;
        this.executingPlayer = executingPlayer;
        this.actionType = actionType;
        this.civilisation = civilisation;
        this.offset = 0;
    }

    public static from(source: PlayerEvent): PlayerEvent {
        const playerEvent = new PlayerEvent(source.player, source.actionType, source.civilisation, source.executingPlayer);
        playerEvent.offset = source.offset;
        return playerEvent;
    }
}

export default PlayerEvent;
