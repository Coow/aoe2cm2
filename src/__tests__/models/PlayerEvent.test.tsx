import PlayerEvent from "../../models/PlayerEvent";

it('PlayerEvent can deserialize old state', () => {
    const source = JSON.parse('{"player": "HOST", "actionType": "PICK", "civilisation": {}, "offset": 123}') as PlayerEvent;
    expect(PlayerEvent.from(source)).toMatchSnapshot();
});

it('PlayerEvent can deserialize new state', () => {
    const source = JSON.parse('{"player": "HOST", "actionType": "PICK", "civilisation": {}, "offset": 123, "executingPlayer": "GUEST"}') as PlayerEvent;
    expect(PlayerEvent.from(source)).toMatchSnapshot();
});
