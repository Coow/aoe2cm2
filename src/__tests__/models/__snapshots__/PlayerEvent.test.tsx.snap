// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerEvent can deserialize new state 1`] = `
PlayerEvent {
  "actionType": "PICK",
  "civilisation": Object {},
  "executingPlayer": "GUEST",
  "offset": 123,
  "player": "HOST",
}
`;

exports[`PlayerEvent can deserialize old state 1`] = `
PlayerEvent {
  "actionType": "PICK",
  "civilisation": Object {},
  "executingPlayer": "HOST",
  "offset": 123,
  "player": "HOST",
}
`;
